### PR TITLE
C++: Range analysis for unsigned AssignMulExpr

### PR DESCRIPTION
**Pull request by jbj**
This PR sits on top of #4076 since its test output depends on that fix. Merge #4076 first.

### CODE REVIEW

**Comment by jbj** [comment]
Started https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1317/

File: cpp/ql/src/semmle/code/cpp/rangeanalysis/SimpleRangeAnalysis.qll [hunk 1]
```diff
@@ -249,6 +255,14 @@ private predicate defDependsOnDef(
     exprDependsOnDef(assignSub.getRValue(), srcDef, srcVar)
   )
   or
+  exists(UnsignedAssignMulExpr assignMul, RangeSsaDefinition nextDef |
+    def = assignMul and
+    assignMul.getLValue() = nextDef.getAUse(v)
+  |
+    defDependsOnDef(nextDef, v, srcDef, srcVar) or
```
Comment (rdmarsh2) [inline cpp/ql/src/semmle/code/cpp/rangeanalysis/SimpleRangeAnalysis.qll:262]: I'm confused about why we need this recursive call (both here and for the other `AssignExpr` subclasses).

Comment (jbj) [reply cpp/ql/src/semmle/code/cpp/rangeanalysis/SimpleRangeAnalysis.qll]: Well spotted. I don't think the desugared version (`x = x * e`) does this recursion. I've pushed a commit to remove the recursion (but still take dependency information from the LHS).

Comment (jbj) [reply cpp/ql/src/semmle/code/cpp/rangeanalysis/SimpleRangeAnalysis.qll]: I've started https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1321/ for just this change (edc5e5f).

**Comment by jbj** [comment]
The CPP-Differences jobs LGTM, so I think this is good to merge. [The first job](https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1320) shows only changes that can be attributed to #4076, and [the second job](https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1329) shows no changes. Neither shows any performance difference other than noise.

**Pull request merged**